### DISCUSSION TITLE
Tab completion behaviour (cycling options vs list options)

### DISCUSSION THREAD

**Discussion by eggbean** [chrisant996/clink#566] [Q&A]
I used to have this behaviour: I would type a command, like `git` and after that repeated TABs would cycle through the different subcommands or arguments.

I don't know what I did to change this, but for the last few weeks I get a zsh-like list of arguments with help text. It's great. I didn't know something like this was possible in cmd/clink.

But how was this change made? I did add clink-gizmos and made other changes around the time this started to happen.

While this new behviour is good, I also miss being able to cycle through the options as well. Is this possible? Is it possible to have the list come up the first time I type TAB and then repeated TABs cycle through the options? It's annoying that partial completions don't work.

**Comment by chrisant996** [comment]
Same as earlier questions about it.

Direct link to the docs I mentioned before:

https://chrisant996.github.io/clink/clink.html#gettingstarted_defaultbindings

**Comment by eggbean** [comment]
I did search to see if there were any issues that covered this.

I still have `"\t": old-menu-complete` in my `.inputrc`, so that is why it cycled, but it seems that it was due to using `clink-gizmos` which has changed the behaviour, as its `"\t": "luafunc:fzf_complete"` setting seems to default to `complete` if there is no `**`.

There doesn't seem to be any other setting for this other than `old-menu-complete` and `complete`. Is it not technically possible to have a hybrid of the two? ie. First tab brings up the list and subsequent TABs cycle through the options.

**Reply by chrisant996** [reply]
> I did search to see if there were any issues that covered this.

It's documented features.  The Getting Started section of the docs is useful.

> I still have `"\t": old-menu-complete` in my `.inputrc`, so that is why it cycled, but it seems that it was due to using `clink-gizmos` which has changed the behaviour, as its `"\t": "luafunc:fzf_complete"` setting seems to default to `complete` if there is no `**`.

You turned on fzf.  So now completion works how fzf is designed to work.

Don't want that?  Configure the key bindings differently.

> There doesn't seem to be any other setting for this other than `old-menu-complete` and `complete`.

I think you're looking for settings.

Instead look for Bindable Commands and set key bindings to whatever you want.  The documentation about key bindings would be useful to read.  There are unlimited ways to configure key bindings; settings would only allow a couple of ways, so you're looking in the wrong place.

> Is it not technically possible to have a hybrid of the two? ie. First tab brings up the list and subsequent TABs cycle through the options.

Read the docs.  What you want is covered.  `clink-select-complete`.

I cannot emphasize enough:  reading the documentation will be useful, especially the Getting Started section.  It sounds like you're making a bunch of changes without reading what the changes will do.  That's fine!  Experiment and enjoy!  But to find out what's happening and how to configure things, please consider reading the documentation.
